### PR TITLE
fix(theme): add separate header and navigation color controls

### LINKED ISSUE #2807: Improve color controls to prevent unreadable header and navigation when bright primary color is selected
When a bright **primary color** is selected for an event, the system currently applies the same color to multiple UI elements. This can create readability issues.

Specifically, the chosen color affects:

* The **top header background**
* The **event date/time text**
* The **navigation menu text**

If a bright color is chosen, the contrast between the background and the text may become too low, making menu items and other information difficult to read.

Image
Image

### Current behavior

* The primary color is applied to several UI elements without sufficient contrast control.
* Menu items can become unreadable when the same bright color is used.
* Event title and date/time may also become difficult to read depending on the selected color.

### Expected behavior

The system should provide **separate color controls** so organisers can maintain proper contrast while still customizing the event appearance.

Color configuration should include:

* **Header background color** (this color should also be used for the **event date/time text background area** so these remain visually aligned)
* **Header text color** (event title and date/time text)
* **Navigation menu text color**
* **Primary accent color** (buttons, highlights)

The **top header background and the area containing the event date/time should continue to share the same background color** so the header remains visually consistent.

### Possible improvements

Optional enhancements:

* Implement automatic **contrast detection** (e.g. WCAG contrast ratio checks).
* If the chosen color creates insufficient contrast:

  * Automatically switch text color to black or white.
  * Show a warning in the color picker.

### Acceptance criteria

* Header background color and the background of the event date/time area use the same color.
* Header text color (event title and date/time text) can be configured independently.
* Navigation menu text color can be configured independently.
* Bright color choices no longer make navigation or header information unreadable.
* The interface maintains acceptable contrast across all pages.

### PR DESCRIPTION
**Pull request by pi-dal**
## Summary
This PR adds separate theme controls for the public event header so organizers can avoid unreadable combinations when using a bright primary color.

## What changed
- added new design settings for:
  - `Header background color`
  - `Header text color`
  - `Navigation text color`
- exposed the new settings in both organizer-level and event-level design forms
- propagated organizer defaults to event settings where appropriate
- updated the generated widget CSS so these values are exported as dedicated CSS variables
- updated CFP and presale header styling to use the new variables while keeping the primary color for accent and active states
- updated the color ETag generation so style changes invalidate the cached CSS correctly
- added focused regression coverage for:
  - generated CSS output
  - cache invalidation when header-related colors change
  - presence of the new settings fields in the event settings form

## Why this change
Issue #2807 describes a readability problem where the primary color was reused for multiple header elements. When a bright color was selected, the header background, event title/date area, and navigation text could become hard to read.

This change separates those responsibilities so organizers can style:
- the header background independently
- the header text independently
- the navigation text independently
- while still keeping the primary color for accents such as highlights and active states

## Validation
- manually verified the new settings in the organizer and event design pages
- manually verified the public event page after saving the new colors:
  - header background color applied correctly
  - header text color applied correctly
  - navigation text color applied correctly
- confirmed the issue-specific UI behavior in the local Docker development environment

## Scope
This PR intentionally includes only the issue-related theme changes and their focused tests. Local Docker helper changes used during validation were not included in the commit.

Closes #2807
Related context: https://github.com/fossasia/eventyay/issues/2807#issuecomment-4083803496

### CODE REVIEW

**Comment by pi-dal** [comment]
## Screenshots
Settings
Public Page


### Before
- Header and navigation styling depended too heavily on the primary color, which could reduce readability.

### After
- Header background color can be configured independently.
- Header text color can be configured independently.
- Navigation text color can be configured independently.
- Primary color remains available for accent and active states.

### Settings
- Added separate controls for:
  - `Header background color`
  - `Header text color`
  - `Navigation text color`

File: app/eventyay/agenda/views/widget.py [hunk 1]
```diff
@@ -18,7 +18,14 @@
...
 def color_etag(request, organizer=None, event=None, **kwargs):
-    return request.event.visible_primary_color or 'none'
+    return '|'.join(
+        [
+            request.event.visible_primary_color or '',
+            request.event.settings.get('header_background_color') or '',
+            request.event.settings.get('header_text_color') or '',
+            request.event.settings.get('navigation_text_color') or '',
+        ]
+    ) or 'none'
```
[the remark this answers is not in the document]
Comment (pi-dal) [reply app/eventyay/agenda/views/widget.py]: Fixed. color_etag() now returns 'none' when no relevant color value is configured, instead of relying on the truthiness of the joined string. I also added a focused regression test for the empty-color case.

**Comment by atrip0305** [comment]
Hi, this looks like a well-structured solution — especially the introduction of separate color roles and CSS variables.

I was exploring this issue as well and noticed that while color separation is handled, low-contrast combinations are still possible (e.g., bright backgrounds with light text).

I’d like to work on an enhancement:
- Add WCAG-based contrast validation for header and navigation colors
- Provide a warning when contrast is insufficient
- Optionally auto-adjust text color (black/white) for better readability

Please let me know if this would be a useful addition.

**Comment by pi-dal** [comment]
> Hi, this looks like a well-structured solution — especially the introduction of separate color roles and CSS variables.
> 
> I was exploring this issue as well and noticed that while color separation is handled, low-contrast combinations are still possible (e.g., bright backgrounds with light text).
> 
> I’d like to work on an enhancement:
> 
> Add WCAG-based contrast validation for header and navigation colors
> Provide a warning when contrast is insufficient
> Optionally auto-adjust text color (black/white) for better readability
> Please let me know if this would be a useful addition.

Hi, thanks for the suggestion — that’s a valid point.

I agree this would be a useful accessibility improvement. However, I’d prefer to keep this PR focused and minimal, limited to introducing the color roles and CSS variables.

Let’s handle contrast validation as a separate follow-up (issue/PR). Happy to review it.

**Review by ArnavBallinCode — CHANGES_REQUESTED** [review]
Thank you for your contribution, please resolve merge conflicts

**Comment by pi-dal** [comment]
> Thank you for your contribution, please resolve merge conflicts

Thank you for the review. I have resolved the merge conflicts and updated the PR with the latest changes from `dev`. Please take another look when you have a moment.

**Comment by Rachit7168** [comment]
Please resolve merge conflicts

**Comment by pi-dal** [comment]
Thank you for the review. I have resolved the merge conflicts and updated the PR with the latest changes. Please take another look when you have a moment.

**Comment by prayag78** [comment]
Please resolve Copilot comments

File: app/eventyay/base/configurations/default_setting.py [hunk 2]
```diff
@@ -2576,6 +2648,9 @@ def primary_font_kwargs():
 
 CSS_SETTINGS = {
     'primary_color',
+    'header_background_color',
+    'header_text_color',
+    'navigation_text_color',
     'theme_color_success',
     'theme_color_danger',
     'primary_font',
```
[the remark this answers is not in the document]
Comment (pi-dal) [reply app/eventyay/base/configurations/default_setting.py:2558]: Fixed. Added `header_background_color`, `header_text_color`, and `navigation_text_color` to both settings serializers so the event and organizer settings APIs expose the new fields for read/write. Also added regression assertions in the existing event and organizer settings API tests to cover the new keys.

**Comment by pi-dal** [comment]
Resolved the Copilot comment and pushed the fix. The new header color settings are now exposed through both the event and organizer settings APIs, with regression coverage added for those fields. Please take another look.

File: app/eventyay/agenda/views/widget.py [hunk 3]
```diff
@@ -264,14 +270,25 @@ def widget_script(request, organizer=None, event=None, **kwargs):
...
+    if request.event.settings.get('header_background_color'):
+        variables.append(
+            f'--color-header-background: {request.event.settings.get("header_background_color")};'
+        )
+    if request.event.settings.get('header_text_color'):
+        variables.append(f'--color-header-text: {request.event.settings.get("header_text_color")};')
+    if request.event.settings.get('navigation_text_color'):
+        variables.append(
+            f'--color-header-navigation: {request.event.settings.get("navigation_text_color")};'
+        )
```
[the remark this answers is not in the document]
Comment (MukundC25) [reply app/eventyay/agenda/views/widget.py]: @pi-dal consider this, a minor suggestion but worth addressing.

[the remark this answers is not in the document]
Comment (pi-dal) [reply app/eventyay/agenda/views/widget.py]: Fixed. I pulled the three header-related settings into local variables in `event_css()`, so each value is read once while the generated CSS stays unchanged. Please have another look when convenient.

**Comment by pi-dal** [comment]
@Saksham-Sirohi @MukundC25 the latest review follow-up is addressed and the branch is up to date with `dev`. Please take another look when convenient.

**Comment by Rachit7168** [comment]
@codex review

File: app/eventyay/base/configurations/default_setting.py [hunk 4]
```diff
@@ -2024,6 +2024,78 @@ def primary_font_kwargs():
...
+            label=_('Navigation text color'),
+            validators=[
+                RegexValidator(
+                    regex='^#[0-9a-fA-F]{6}$',
+                    message=_('Please enter the hexadecimal code of a color, e.g. #990000.'),
+                ),
+            ],
+            widget=forms.TextInput(attrs={'class': 'colorpickerfield'}),
+        ),
+    },
```
[the remark this answers is not in the document]
Comment (sarafarajnasardi) [reply app/eventyay/base/configurations/default_setting.py]: ptal @pi-dal

**Review by sarafarajnasardi — CHANGES_REQUESTED** [review]
Please address the copilot reviews

File: app/eventyay/base/configurations/default_setting.py [hunk 5]
```diff
@@ -2024,6 +2024,78 @@ def primary_font_kwargs():
             widget=forms.TextInput(attrs={'class': 'colorpickerfield'}),
         ),
     },
+    'header_background_color': {
```
Comment (SxxAq) [inline app/eventyay/base/configurations/default_setting.py]: **CODE QUALITY** 

Please reduce repetition in color field definitions.

4 color fields (primary_color, header_background_color, 
header_text_color, navigation_text_color) with identical validator patterns.

The same regex validator appears in all 4 settings:
  regex='^#[0-9a-fA-F]{6}$'
  message=_('Please enter the hexadecimal code of a color, e.g. #990000.')

This ~80 lines of code could be reduced to ~20 lines using a factory function:

Suggested approach:
```python
def _hex_color_field_config(label):
    hex_validator = RegexValidator(
        regex='^#[0-9a-fA-F]{6}$',
        message=_('Please enter the hexadecimal code of a color, e.g. #990000.'),
    )
    return {
        'default': '',
        'type': str,
        'form_class': forms.CharField,
        'serializer_class': serializers.CharField,
        'serializer_kwargs': dict(validators=[hex_validator]),
        'form_kwargs': dict(
            label=label,
            validators=[hex_validator],
            widget=forms.TextInput(attrs={'class': 'colorpickerfield'}),
        ),
    }

# Then use:
'primary_color': _hex_color_field_config(_('Primary color')),
'header_background_color': _hex_color_field_config(_('Header background color')),
'header_text_color': _hex_color_field_config(_('Header text color')),
'navigation_text_color': _hex_color_field_config(_('Navigation text color')),

File: app/eventyay/eventyay_common/forms/event.py
Comment (SxxAq) [inline app/eventyay/eventyay_common/forms/event.py:96]: **Missing validation for header color pairs**

The clean() method validates image fields but NOT the new color settings.
If header_background_color is set without header_text_color, 
the text defaults to white (from CSS variables), creating unreadable contrast.

eg: 
- Organizer sets: header_background_color = '#ffff00' (bright yellow)
- Organizer leaves: header_text_color = '' (empty, defaults to white)
- Result: White text on yellow = UNREADABLE (same problem the PR tries to solve!)

REQUIRED FIX: Add validation in clean() to require both colors together.

**Review by SxxAq — CHANGES_REQUESTED** [review]

File: app/eventyay/base/configurations/default_setting.py (hunk 5 above)
Comment (pi-dal) [reply app/eventyay/base/configurations/default_setting.py]: Addressed. I extracted the repeated hex color field wiring into a shared helper and switched the header/theme color settings to use it, which removes the validator/widget duplication without changing the field metadata.

File: app/tests/talk/agenda/views/test_agenda_widget.py [hunk 6]
```diff
@@ -87,3 +119,55 @@ def test_anon_cannot_access_wip_schedule(client, event):
...
+    request.event = make_event(header_background_color='#ffee00')
+
+    assert color_etag(request) != default_etag
+
+
+def test_event_css_etag_returns_none_when_no_colors_are_set(rf):
+    request = rf.get('/settings.css')
+    request.event = make_event(primary_color='')
+
+    assert color_etag(request) == 'none'
```
[the remark this answers is not in the document]
Comment (pi-dal) [reply app/tests/talk/agenda/views/test_agenda_widget.py]: Renamed the test to make it explicit that the fallback value is the string sentinel `"none"`, not Python `None`.

File: app/eventyay/agenda/views/widget.py [hunk 7]
```diff
@@ -27,7 +27,13 @@
...
 
 def color_etag(request, organizer=None, event=None, **kwargs):
-    return request.event.visible_primary_color or 'none'
+    parts = [
+        request.event.visible_primary_color or '',
+        request.event.settings.get('header_background_color') or '',
+        request.event.settings.get('header_text_color') or '',
+        request.event.settings.get('navigation_text_color') or '',
+    ]
+    return '|'.join(parts) if any(parts) else 'none'
```
[the remark this answers is not in the document]
Comment (pi-dal) [reply app/eventyay/agenda/views/widget.py:39]: Addressed. `color_etag()` now mirrors `event_css()` by reading the header color settings into locals before building the ETag parts.

File: app/eventyay/eventyay_common/forms/event.py
Comment (pi-dal) [reply app/eventyay/eventyay_common/forms/event.py:96]: I checked this one and I am not applying the pair-validation constraint. The current CSS fallback is intentionally what makes these roles independently overridable: leaving `header_text_color` or `navigation_text_color` empty falls back to the existing header defaults from `app/eventyay/static/common/css/_variables.css`. Requiring both fields together would change that behavior and narrow the PR from independent color roles into a new UX validation policy. The contrast/readability concern is real, but it needs a dedicated WCAG-style validation/fallback follow-up rather than coupling these fields at form-clean time.

File: app/eventyay/agenda/views/widget.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pi-dal) [reply app/eventyay/agenda/views/widget.py]: Addressed. `event_css()` now reads the header color settings into locals before building the CSS variables.

**Review by Rachit7168 — APPROVED** [review]
LGTM , Tested 

https://github.com/user-attachments/assets/65cb93f9-fe10-4133-b791-f3085708ab38

**Review by ArnavBallinCode — COMMENTED** [review]
Can you confirm whether the colour picker warnings check the colours of header background and header/navigation text colours, or if they only validate individual colours against the default background?

**Comment by pi-dal** [comment]
@ArnavBallinCode Confirmed: the current color picker warning is the existing per-field contrast check. These new header color fields use `colorpickerfield`, so they get that warning, but the check compares the selected color against a white/default background. It does not currently validate cross-field combinations such as header text or navigation text against the custom header background.

If we want pairwise header contrast validation, that should be handled as a separate enhancement.

**Review by sarafarajnasardi — APPROVED** [review]
LGTM

**Review by ArnavBallinCode — APPROVED** [review]
LGTM, please open a follow up post merge.

**Review by SxxAq — APPROVED** [review]
LGTM

**Review by Saksham-Sirohi — APPROVED** [review]
